### PR TITLE
Prefer rackup file specified by CLI

### LINKED ISSUE #2225: Rackup overrides CLI preference for app
**Describe the bug**

Puma should always prefer the rackup file specified by the CLI, but if the `rackup` DSL command is used, it does not.

Related: #1401

**Puma config:**

Using the test fixtures in the repo: 

```
rackup "test/rackup/hello-env.ru"
```

**To Reproduce**

```
bin/puma -C my_rackup_config.rb test/rackup/hello.ru
```

...should return a Hello World! response, but runs the hello-env app instead.

### PR DESCRIPTION
**Pull request by acmh**
### Description
This PR aims to always give preference for the rackup file specified by the CLI. I also took opportunity and added some tests.

This Closes #2225 .

### Your checklist for this pull request

- [x] I have reviewed the [guidelines for contributing](../blob/master/CONTRIBUTING.md) to this repository.
- [x] I have added an entry to [History.md](../blob/master/History.md) if this PR fixes a bug or adds a feature. If it doesn't need an entry to HISTORY.md, I have added `[changelog skip]` the pull request title.
- [x] I have added appropriate tests if this PR fixes a bug or adds a feature.
- [x] My pull request is 100 lines added/removed or less so that it can be easily reviewed.
- [ ] If this PR doesn't need tests (docs change), I added `[ci skip]` to the title of the PR.
- [x] If this closes any issues, I have added "Closes `#issue`" to the PR description or my commit messages.
- [ ] I have updated the documentation accordingly.
- [x] All new and existing tests passed, including Rubocop.

### CODE REVIEW

**Comment by acmh** [comment]
cc @nateberkopec

File: test/test_integration_single.rb [hunk 1]
```diff
@@ -43,6 +43,19 @@ def test_term_suppress
     assert_equal 0, status
   end
 
+  def test_prefer_rackup_file_specified_by_cli
+    skip_unless_signal_exist? :TERM
+    skip_on :jruby
```
Comment (nateberkopec) [inline test/test_integration_single.rb:48]: You can remove this line

File: test/test_integration_single.rb [hunk 2]
```diff
@@ -43,6 +43,19 @@ def test_term_suppress
...
+
+    cli_server "-C test/config/with_rackup_from_dsl.rb test/rackup/hello.ru"
+    connection = connect
+    reply = read_body(connection)
+    _, status = stop_server
+
+    assert_match("Hello World", reply)
+    assert_equal 15, status
```
Comment (nateberkopec) [inline test/test_integration_single.rb:56]: This assertion isn't necessary I think, we only need the correct reply.

**Comment by nateberkopec** [comment]
Thank you so much!

Can you confirm this new test fails on master?

**Comment by acmh** [comment]
All changes requested were made.

> Thank you so much!
> 
> Can you confirm this new test fails on master?

Yes, I confirm.

File: lib/puma/dsl.rb [hunk 3]
```diff
@@ -308,7 +308,7 @@ def debug
     # @example
     #   rackup '/u/apps/lolcat/config.ru'
     def rackup(path)
-      @options[:rackup] = path.to_s
+      @options[:rackup] = path.to_s unless @options[:rackup]
```
Comment (cristian-rivera) [inline lib/puma/dsl.rb:311]:
```suggestion
      @options[:rackup] ||= path.to_s
```
Perhaps even simpler?

Comment (acmh) [reply lib/puma/dsl.rb:311]: @cristian-rivera I've tried this approach on [26393ae](https://github.com/puma/puma/pull/2236/commits/26393ae933dcad90510348db9033b14a67075e79) but the tests failed on two builds, so I reverted the change.

Comment (nateberkopec) [reply lib/puma/dsl.rb:311]: These suggested changes -> commit things are so nice.

**Comment by rmachielse** [comment]
Is it possible to backport this to 4.3? We are experiencing this issue with puma 4.3 too.

**Pull request merged**